rclcpp/test/rclcpp/test_publisher.cpp

```diff
@@ -259,3 +264,178 @@ TEST_F(TestPublisher, basic_getters) {
+  auto publisher = node->create_publisher<test_msgs::msg::Empty>("topic", 10);
+  RCLCPP_EXPECT_THROW_EQ(
+    publisher->get_actual_qos(),
+    std::runtime_error("failed to get qos settings: error not set"));
+}
+
+TEST_F(TestPublisher, publishers_equal_rmw_compare_gids_error) {
+  initialize();
+  auto mock = mocking_utils::patch_and_return(
+    "lib:rclcpp", rmw_compare_gids_equal, RMW_RET_ERROR);
+  const auto publisher = node->create_publisher<test_msgs::msg::Empty>("topic", 10);
+  const rmw_gid_t * gid = nullptr;
+  auto throwing_fn = [publisher, gid]()
+    {
+      // The == operator is expected to throw here, but this lambda avoids unused result warning
+      return (*publisher.get() == gid) ? true : false;
+    };
+
+  RCLCPP_EXPECT_THROW_EQ(
+    throwing_fn(),
+    std::runtime_error("failed to compare gids: error not set"));
+}
+
+TEST_F(TestPublisher, intra_process_publish_failures) {
+  initialize();
+  rclcpp::PublisherOptionsWithAllocator<std::allocator<void>> options;
+  options.use_intra_process_comm = rclcpp::IntraProcessSetting::Enable;
+  auto publisher = node->create_publisher<test_msgs::msg::Empty>("topic", 10, options);
+
+  auto msg_unique = std::make_unique<test_msgs::msg::Empty>();
+  EXPECT_NO_THROW(publisher->publish(std::move(msg_unique)));
+
+  rclcpp::SerializedMessage serialized_msg;
+  RCLCPP_EXPECT_THROW_EQ(
+    publisher->publish(serialized_msg),
+    std::runtime_error("storing serialized messages in intra process is not supported yet"));
+
+  std::allocator<void> allocator;
+  {
+    rclcpp::LoanedMessage<test_msgs::msg::Empty> loaned_msg(*publisher, allocator);
+    RCLCPP_EXPECT_THROW_EQ(
+      publisher->publish(std::move(loaned_msg)),
+      std::runtime_error("storing loaned messages in intra process is not supported yet"));
+  }
+
+  {
+    rclcpp::LoanedMessage<test_msgs::msg::Empty> loaned_msg(*publisher, allocator);
+    loaned_msg.release();
+    RCLCPP_EXPECT_THROW_EQ(
+      publisher->publish(std::move(loaned_msg)),
+      std::runtime_error("loaned message is not valid"));
+  }
+}
+
+TEST_F(TestPublisher, inter_process_publish_failures) {
+  initialize();
+  rclcpp::PublisherOptionsWithAllocator<std::allocator<void>> options;
+  options.use_intra_process_comm = rclcpp::IntraProcessSetting::Disable;
+  auto publisher = node->create_publisher<test_msgs::msg::Empty>("topic", 10, options);
+
+  auto msg_unique = std::make_unique<test_msgs::msg::Empty>();
+  EXPECT_NO_THROW(publisher->publish(std::move(msg_unique)));
+
+  {
+    auto mock = mocking_utils::patch_and_return(
+      "lib:rclcpp", rcl_publish, RCL_RET_PUBLISHER_INVALID);
+    test_msgs::msg::Empty msg;
+    EXPECT_THROW(publisher->publish(msg), rclcpp::exceptions::RCLError);
+  }
+
+  rclcpp::SerializedMessage serialized_msg;
+  EXPECT_NO_THROW(publisher->publish(serialized_msg));
+
+  {
+    auto mock = mocking_utils::patch_and_return(
+      "self", rcl_publish_serialized_message, RCL_RET_ERROR);
```

Comment:
@brawner nit: mind to add a note explaining why `self` is used here instead of `lib:rclcpp`? It may be helpful to other people.

Reply:
Done